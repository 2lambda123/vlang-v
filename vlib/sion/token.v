enum TokenKind{
	eof
	nil
	date
	data
	integer
	double
	str
}