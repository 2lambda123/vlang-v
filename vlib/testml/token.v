module

enum TokenKind{
	eof
	name
	function
	str
	hyphen
	json
	equal
	yaml
}

fn build_token_str() []string{

}