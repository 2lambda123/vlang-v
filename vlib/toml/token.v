import toml

enum Token{
    nil
    eof
    text
    str
    raw_str
    multiline_string
    boolean
    integer
    float
    datetime
    array_start
    array_end
    table_start
    table_end
    table_sep
    key_sep
    array_table_start
    array_table_end
    key_start
}

fn build_token_str() []string{
    mut s := ['', NrTokens]
    s[Token.eof] = 'eof'
    s[Token.table_start] = '['
    s[Token.table_end] = ']'
    s[Token.array_table_start] = '['
    s[Token.array_table_end] = ']'
}