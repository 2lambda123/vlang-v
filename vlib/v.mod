#V Library#

Module {
	name: 'vlib',
	description: 'The V standart library, provided by V itself.',
	dependencies: []
}
