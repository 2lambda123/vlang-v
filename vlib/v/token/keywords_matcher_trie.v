module token

pub struct TrieNode {
mut:
	children       [256]&TrieNode
	is_end_of_word bool
	kind           Kind
	prefix         string
	min_len        int = 999999
	max_len        int
}

pub fn new_trie_node(prefix string) &TrieNode {
	return &TrieNode{
		prefix: prefix
	}
}

pub fn (node &TrieNode) show(level int) {
	mut non_nil_children := 0
	for x in node.children {
		if x != unsafe { nil } {
			non_nil_children++
		}
	}
	eprintln('> level: ${level:2} | prefix: ${node.prefix:20} | kind: ${node.kind:12} | minl: ${node.min_len:2} | maxl: ${node.max_len:2} | is_end_of_word: ${int(node.is_end_of_word)} | non_nil_children: ${non_nil_children:2}')
	for x in node.children {
		if x != unsafe { nil } {
			x.show(level + 1)
		}
	}
}

pub fn (mut node TrieNode) add_word(word string, kind Kind, word_idx int) {
	if node.max_len < word.len {
		node.max_len = word.len
	}
	if node.min_len > word.len {
		node.min_len = word.len
	}
	first := u8(word[word_idx] or {
		node.is_end_of_word = true
		node.kind = kind
		return
	})
	// eprintln('>> node: ${ptr_str(node)} | first: $first | word_idx: $word_idx')
	mut child_node := node.children[first]
	if child_node == unsafe { nil } {
		child_node = new_trie_node(word#[..word_idx + 1])
		node.children[first] = child_node
	}
	child_node.add_word(word, kind, word_idx + 1)
}

[direct_array_access]
pub fn (root &TrieNode) find(word string) int {
	if word.len > root.max_len || word.len < root.min_len {
		return -1
	}
	mut node := unsafe { &TrieNode(root) }
	mut idx := 0
	for {
		// eprintln('> match_keyword: `${word:20}` | node: ${ptr_str(node)} | node.prefix: ${node.prefix:15} | idx: ${idx:3}')
		if idx == word.len {
			if node.is_end_of_word {
				// node.show(0)
				return int(node.kind)
			}
			return -1
		}
		c := word[idx]
		child := node.children[c]
		if child == unsafe { nil } {
			return -1
		}
		node = child
		idx++
	}
	return -1
}

[heap]
pub struct KeywordsMatcherTrie {
mut:
	nodes   []&TrieNode
	min_len int = 999999
	max_len int
}

[direct_array_access]
pub fn (km &KeywordsMatcherTrie) find(word string) int {
	wlen := word.len
	if wlen < km.min_len {
		return -1
	}
	if wlen > km.max_len {
		return -1
	}
	node := km.nodes[wlen]
	if node == unsafe { nil } {
		return -1
	}
	return node.find(word)
}

[direct_array_access]
pub fn (mut km KeywordsMatcherTrie) add_word(word string, kind Kind) {
	wlen := word.len
	if km.max_len < wlen {
		km.max_len = wlen
	}
	if km.min_len > wlen {
		km.min_len = wlen
	}
	if km.nodes[wlen] == unsafe { nil } {
		km.nodes[wlen] = new_trie_node('')
	}
	km.nodes[wlen].add_word(word, kind, 0)
}

pub fn new_keywords_matcher_trie<T>(kw_map map[string]T) KeywordsMatcherTrie {
	mut km := KeywordsMatcherTrie{
		nodes: []&TrieNode{cap: 20}
	}
	for _ in 0 .. 20 {
		km.nodes << &TrieNode(0)
	}
	for k, v in kw_map {
		km.add_word(k, v)
	}
	// dump(km.min_len)
	// dump(km.max_len)
	// for idx,x in km.nodes { if x != unsafe { nil } { eprintln('>> idx: $idx | ${ptr_str(x)}') } }
	return km
}

pub fn new_keywords_matcher_from_array_trie(names []string) KeywordsMatcherTrie {
	mut m := map[string]int{}
	for i, name in names {
		m[name] = i
	}
	return new_keywords_matcher_trie<int>(m)
}
