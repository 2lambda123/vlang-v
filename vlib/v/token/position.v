// Copyright (c) 2019-2020 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module token

pub struct Position {
pub:
	line_nr int // the line number in the source where the token occured
	pos     int // the position of the token in scanner text
	len     int // length of the literal
}

[inline]
pub fn (tok &Token) position() Position {
	return Position{
		line_nr: tok.line_nr - 1
		pos: tok.pos
		len: tok.len
	}
}
