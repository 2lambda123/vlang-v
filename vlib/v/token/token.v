// Copyright (c) 2019 Alexander Medvednikov. All rights reserved.
// Use of this source code is governed by an MIT license
// that can be found in the LICENSE file.
module token

pub struct Token {
pub:
	kind     TokenKind // the token number/enum; for quick comparisons
	lit      string // literal representation of the token
	// line_nr  int // the line number in the source where the token occured
	// name_idx int // name table index for O(1) lookup
	// pos      int // the position of the token in scanner text
}

pub enum TokenKind {
	eof
	name // user
	number // 123
	str // 'foo'
	str_inter // 'name=$user.name'
	chartoken // `A`
	plus
	minus
	mul
	div
	mod
	xor // ^
	pipe // |
	inc // ++
	dec // --
	and // &&
	logical_or
	not
	bit_not
	question
	comma
	semicolon
	colon
	arrow // =>
	amp
	hash
	dollar
	str_dollar
	left_shift
	righ_shift
	// at // @
	assign // =
	decl_assign // :=
	plus_assign // +=
	minus_assign // -=
	div_assign
	mult_assign
	xor_assign
	mod_assign
	or_assign
	and_assign
	righ_shift_assign
	left_shift_assign
	// {}  () []
	lcbr
	rcbr
	lpar
	rpar
	lsbr
	rsbr
	// == != <= < >= >
	eq
	ne
	gt
	lt
	ge
	le
	// comments
	line_comment
	mline_comment
	nl
	dot
	dotdot
	ellipsis
	// keywords
	keyword_beg
	key_as
	key_asm
	key_assert
	key_atomic
	key_break
	key_const
	key_continue
	key_defer
	key_else
	key_embed
	key_enum
	key_false
	key_for
	key_fn
	key_global
	key_go
	key_goto
	key_if
	key_import
	key_import_const
	key_in
	key_interface
	// key_it
	key_match
	key_module
	key_mut
	key_none
	key_return
	key_select
	key_sizeof
	key_offsetof
	key_struct
	key_switch
	key_true
	key_type
	// typeof
	key_orelse
	key_union
	key_pub
	key_static
	key_unsafe
	keyword_end
}

const (
	assign_tokens = [TokenKind.assign, .plus_assign, .minus_assign, .mult_assign,
		.div_assign, .xor_assign, .mod_assign, .or_assign, .and_assign,
		.righ_shift_assign, .left_shift_assign]

	nr_tokens = 141
)


// build_keys genereates a map with keywords' string values:
// Keywords['return'] == .key_return
fn build_keys() map[string]int {
	mut res := map[string]int
	for t := int(TokenKind.keyword_beg) + 1; t < int(TokenKind.keyword_end); t++ {
		key := token_str[t]
		res[key] = t
	}
	return res
}

// TODO remove once we have `enum TokenKind { name('name') if('if') ... }`
fn build_token_str() []string {
	mut s := [''].repeat(nr_tokens)
	s[TokenKind.keyword_beg] = ''
	s[TokenKind.keyword_end] = ''
	s[TokenKind.eof] = 'eof'
	s[TokenKind.name] = 'name'
	s[TokenKind.number] = 'number'
	s[TokenKind.str] = 'STR'
	s[TokenKind.chartoken] = 'char'
	s[TokenKind.plus] = '+'
	s[TokenKind.minus] = '-'
	s[TokenKind.mul] = '*'
	s[TokenKind.div] = '/'
	s[TokenKind.mod] = '%'
	s[TokenKind.xor] = '^'
	s[TokenKind.bit_not] = '~'
	s[TokenKind.pipe] = '|'
	s[TokenKind.hash] = '#'
	s[TokenKind.amp] = '&'
	s[TokenKind.inc] = '++'
	s[TokenKind.dec] = '--'
	s[TokenKind.and] = '&&'
	s[TokenKind.logical_or] = '||'
	s[TokenKind.not] = '!'
	s[TokenKind.dot] = '.'
	s[TokenKind.dotdot] = '..'
	s[TokenKind.ellipsis] = '...'
	s[TokenKind.comma] = ','
	// s[TokenKind.at] = '@'
	s[TokenKind.semicolon] = ';'
	s[TokenKind.colon] = ':'
	s[TokenKind.arrow] = '=>'
	s[TokenKind.assign] = '='
	s[TokenKind.decl_assign] = ':='
	s[TokenKind.plus_assign] = '+='
	s[TokenKind.minus_assign] = '-='
	s[TokenKind.mult_assign] = '*='
	s[TokenKind.div_assign] = '/='
	s[TokenKind.xor_assign] = '^='
	s[TokenKind.mod_assign] = '%='
	s[TokenKind.or_assign] = '|='
	s[TokenKind.and_assign] = '&='
	s[TokenKind.righ_shift_assign] = '>>='
	s[TokenKind.left_shift_assign] = '<<='
	s[TokenKind.lcbr] = '{'
	s[TokenKind.rcbr] = '}'
	s[TokenKind.lpar] = '('
	s[TokenKind.rpar] = ')'
	s[TokenKind.lsbr] = '['
	s[TokenKind.rsbr] = ']'
	s[TokenKind.eq] = '=='
	s[TokenKind.ne] = '!='
	s[TokenKind.gt] = '>'
	s[TokenKind.lt] = '<'
	s[TokenKind.ge] = '>='
	s[TokenKind.le] = '<='
	s[TokenKind.question] = '?'
	s[TokenKind.left_shift] = '<<'
	s[TokenKind.righ_shift] = '>>'
	s[TokenKind.line_comment] = '// line comment'
	s[TokenKind.mline_comment] = '/* mline comment */'
	s[TokenKind.nl] = 'NLL'
	s[TokenKind.dollar] = '$'
	s[TokenKind.str_dollar] = '$2'
	s[TokenKind.key_assert] = 'assert'
	s[TokenKind.key_struct] = 'struct'
	s[TokenKind.key_if] = 'if'
	// s[TokenKind.key_it] = 'it'
	s[TokenKind.key_else] = 'else'
	s[TokenKind.key_asm] = 'asm'
	s[TokenKind.key_return] = 'return'
	s[TokenKind.key_module] = 'module'
	s[TokenKind.key_sizeof] = 'sizeof'
	s[TokenKind.key_go] = 'go'
	s[TokenKind.key_goto] = 'goto'
	s[TokenKind.key_const] = 'const'
	s[TokenKind.key_mut] = 'mut'
	s[TokenKind.key_type] = 'type'
	s[TokenKind.key_for] = 'for'
	s[TokenKind.key_switch] = 'switch'
	s[TokenKind.key_fn] = 'fn'
	s[TokenKind.key_true] = 'true'
	s[TokenKind.key_false] = 'false'
	s[TokenKind.key_continue] = 'continue'
	s[TokenKind.key_break] = 'break'
	s[TokenKind.key_import] = 'import'
	s[TokenKind.key_embed] = 'embed'
	s[TokenKind.key_unsafe] = 'unsafe'
	// TokenKinds[key_typeof] = 'typeof'
	s[TokenKind.key_enum] = 'enum'
	s[TokenKind.key_interface] = 'interface'
	s[TokenKind.key_pub] = 'pub'
	s[TokenKind.key_import_const] = 'import_const'
	s[TokenKind.key_in] = 'in'
	s[TokenKind.key_atomic] = 'atomic'
	s[TokenKind.key_orelse] = 'or'
	s[TokenKind.key_global] = '__global'
	s[TokenKind.key_union] = 'union'
	s[TokenKind.key_static] = 'static'
	s[TokenKind.key_as] = 'as'
	s[TokenKind.key_defer] = 'defer'
	s[TokenKind.key_match] = 'match'
	s[TokenKind.key_select] = 'select'
	s[TokenKind.key_none] = 'none'
	s[TokenKind.key_offsetof] = '__offsetof'
	return s
}

const (
	token_str = build_token_str()
	keywords = build_keys()
)

pub fn key_to_token(key string) TokenKind {
	a := TokenKind(keywords[key])
	return a
}

pub fn is_key(key string) bool {
	return int(key_to_token(key)) > 0
}

pub fn is_decl(t TokenKind) bool {
	return t in [.key_enum,
.key_interface, .key_fn, .key_struct, .key_type, .key_const, .key_import_const,
.key_pub, .eof]
}

fn (t TokenKind) is_assign() bool {
	return t in assign_tokens
}

fn (t []TokenKind) contains(val TokenKind) bool {
	for tt in t {
		if tt == val {
			return true
		}
	}
	return false
}

pub fn (t TokenKind) str() string {
	if t == .number {
		return 'number'
	}
	if t == .chartoken {
		return 'char'//'`lit`'
	}
	if t == .str {
		return 'str' //"'lit'"
	}
	/*
	if t < .plus {
		return lit // string, number etc
	}
	*/
	return token_str[int(t)]
}

pub fn (t Token) str() string {
	return '$t.kind.str() "$t.lit"'
}


// Representation of highest and lowest precedence
pub const (
	lowest_prec  = 0
	highest_prec = 7
)

// Precedence returns a tokens precedence if defined, otherwise lowest_prec
pub fn (tok Token) precedence() int {
	match tok.kind {
		// `*` |  `/` | `%` | `<<` | `>>` | `&`
		.mul, .div, .left_shift, .righ_shift, .amp { return 7 }
		// `+` |  `-` |  `|` | `^`
		.plus, .minus, .pipe, .xor { return 6 }
		// `==` | `!=` | `<` | `<=` | `>` | `>=` | +=
		.eq, .ne, .lt, .le, .gt, .ge, .plus_assign { return 5 }
		// `&&`
		.and { return 4 }
		// `||`
		.logical_or { return 3 }
		else { return lowest_prec }
	}
}

// is_scalar returns true if the token is a scalar
pub fn (tok Token) is_scalar() bool {
	return tok.kind in [.number, .str]
}

// is_unary returns true if the token can be in a unary expression
pub fn (tok Token) is_unary() bool {
	return tok.kind in [
		//  `+` | `-` | `!` | `~` | `*` | `&`
		.plus, .minus, .not, .bit_not, .mul, .amp
	]
}

// NOTE: do we need this for all tokens (is_left_assoc / is_right_assoc),
// or only ones with the same precedence?

// is_left_assoc returns true if the token is left associative
pub fn (tok Token) is_left_assoc() bool {
	return tok.kind in [
		// .number,
		// `*` | `/` | `%`
		.mul, .div, .mod,
		// `^` | `||` | `&`
		.xor, .logical_or, .and,
		// `,`
		.comma
	]
}

// is_right_assoc returns true if the token is right associative
pub fn (tok Token) is_right_assoc() bool {
	return tok.kind in [
		// `+` | `-` | `!` | `++` | `--`
		.plus, .minus, .not, .inc, .dec,
		// `=` | `+=` | `-=` | `*=` | `/=`
		.assign, .plus_assign, .minus_assign, .mult_assign, .div_assign,
		// `%=` | `>>=` | `<<=`
		.mod_assign, .righ_shift_assign, .left_shift_assign,
		// `&=` | `^=` | `|=`
		.and_assign, .xor_assign, .or_assign
	]
}
