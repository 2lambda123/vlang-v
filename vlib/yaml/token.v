module yaml

enum TokenKind{
	eof,
	comma,
	hyphen,
	rangle,
	vertcal,
	hash,
	seprate,
}