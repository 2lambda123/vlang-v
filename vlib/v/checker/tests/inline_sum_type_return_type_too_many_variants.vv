import v.token

fn foo() int | string | token.Position | bool | u32 {
	return 1
}
