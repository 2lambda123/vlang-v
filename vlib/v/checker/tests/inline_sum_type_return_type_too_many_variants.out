vlib/v/checker/tests/inline_sum_type_return_type_too_many_variants.vv:3:10: error: an inline sum type expects a maximum of three types (5 where given)
    1 | import v.token
    2 | 
    3 | fn foo() int | string | token.Position | bool | u32 {
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    4 |     return 1
    5 | }
